### PR TITLE
doc: use TRUC instead of v3 and add release note

### PR DESCRIPTION
**Pull request by glozow**
Adds a release note for TRUC policy which will be live in v28.0.

For clarity, replaces mentions of "v3" with "TRUC" in most places. Suggested in
- https://github.com/bitcoin/bitcoin/pull/29496#discussion_r1629749583
- https://github.com/bitcoin/bitcoin/pull/29496#discussion_r1624500904

I changed error strings from "v3-violation" to "TRUC-violation" but left v3 in the debug strings because I think it might be clearer for somebody who is debugging. Similarly, I left some variables unchanged because I think they're more descriptive this way, e.g. `tx_v3_from_v2_and_v3`. I'm happy to debate places that should or shouldn't be documented differently in this PR, whatever is clearest to everyone.

### CODE REVIEW

**Review by ismaelsadeeq — COMMENTED** [review]
Concept ACK

File: doc/release-notes-29496.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Mempool Policy Changes
+----------------------
+
+- Transactions with version number set to 3 are now treated as standard on all networks (#29496),
+  subject to Opt-in Topologically Restricted Until Confirmation (TRUC) Transactions policy as
+  described in BIP 431.  The policy includes limits on spending unconfirmed outputs (#28948), eviction
```
Comment (instagibbs) [inline doc/release-notes-29496.md:6]: could directly link the BIP here: https://github.com/bitcoin/bips/blob/master/bip-0431.mediawiki

**Review by instagibbs — APPROVED** [review]
ACK a6821a0a4e1cf2aba679962289f4beeefd3d539a

fairly mechanical changes

**Comment by glozow** [comment]
cc @murchandamus @ismaelsadeeq since you suggested the doc changes

File: src/policy/truc_policy.cpp [hunk 2]
```diff
@@ -203,7 +203,7 @@ std::optional<std::pair<std::string, CTransactionRef>> SingleV3Checks(const CTra
 
     // Remaining checks only pertain to transactions with unconfirmed ancestors.
     if (mempool_ancestors.size() > 0) {
-        // If this transaction spends V3 parents, it cannot be too large.
+        // If this transaction spends TRUC parents, it cannot be too large.
         if (vsize > V3_CHILD_MAX_VSIZE) {
             return std::make_pair(strprintf("v3 child tx %s (wtxid=%s) is too big: %u > %u virtual bytes",
```
Comment (ismaelsadeeq) [inline src/policy/truc_policy.cpp:208]: nit: the v3 mentions in debug strings should be explicitly stated as version=3 for clarity.
here and other places
```suggestion
            return std::make_pair(strprintf("version=3 child tx %s (wtxid=%s) is too big: %u > %u virtual bytes",
```

File: src/policy/truc_policy.h [hunk 3]
```diff
@@ -15,11 +15,12 @@
 #include <set>
```
Comment (ismaelsadeeq) [inline src/policy/truc_policy.h:15]: In f5b26e5b9be41cdab6e75aa0a3fbbb765a96ed06
Here are a couple of places that were not replaced
```diff
diff --git a/src/policy/truc_policy.h b/src/policy/truc_policy.h
index d1e65806cc0..37a445bcc14 100644
--- a/src/policy/truc_policy.h
+++ b/src/policy/truc_policy.h
@@ -23,18 +23,18 @@ static constexpr decltype(CTransaction::version) TRUC_VERSION{3};
 // of 2 and ancestor set size of 2.
 /** Maximum number of transactions including an unconfirmed tx and its descendants. */
 static constexpr unsigned int TRUC_DESCENDANT_LIMIT{2};
-/** Maximum number of transactions including a V3 tx and all its mempool ancestors. */
+/** Maximum number of transactions including a TRUC tx and all its mempool ancestors. */
 static constexpr unsigned int TRUC_ANCESTOR_LIMIT{2};
 
-/** Maximum sigop-adjusted virtual size of all v3 transactions. */
+/** Maximum sigop-adjusted virtual size of all TRUC transactions. */
 static constexpr int64_t TRUC_MAX_VSIZE{10000};
-/** Maximum sigop-adjusted virtual size of a tx which spends from an unconfirmed v3 
transaction. */
+/** Maximum sigop-adjusted virtual size of a tx which spends from an unconfirmed TRU
C transaction. */
 static constexpr int64_t TRUC_CHILD_MAX_VSIZE{1000};
 // These limits are within the default ancestor/descendant limits.
 static_assert(TRUC_MAX_VSIZE + TRUC_CHILD_MAX_VSIZE <= DEFAULT_ANCESTOR_SIZE_LIMIT_K
VB * 1000);
 static_assert(TRUC_MAX_VSIZE + TRUC_CHILD_MAX_VSIZE <= DEFAULT_DESCENDANT_SIZE_LIMIT
_KVB * 1000);
 
-/** Must be called for every transaction, even if not v3. Not strictly necessary for
 transactions
+/** Must be called for every transaction, even if not TRUC. Not strictly necessary f
or transactions
  * accepted through AcceptMultipleTransactions.
  *
  * Checks the following rules:
diff --git a/src/test/fuzz/package_eval.cpp b/src/test/fuzz/package_eval.cpp
index ecacfed58ed..71f72847dd4 100644
--- a/src/test/fuzz/package_eval.cpp
+++ b/src/test/fuzz/package_eval.cpp
@@ -225,7 +225,7 @@ FUZZ_TARGET(tx_package_eval, .init = initialize_tx_pool)
                     tx_mut.vin.emplace_back();
                 }
 
-                // Make a p2pk output to make sigops adjusted vsize to violate v3, p
otentially, which is never spent
+                // Make a p2pk output to make sigops adjusted vsize to violate TRUC,
 potentially, which is never spent
                 if (last_tx && amount_in > 1000 && fuzzed_data_provider.ConsumeBool(
)) {
                     tx_mut.vout.emplace_back(1000, CScript() << std::vector<unsigned
 char>(33, 0x02) << OP_CHECKSIG);
                     // Don't add any other outputs.
diff --git a/src/test/util/txmempool.cpp b/src/test/util/txmempool.cpp
index 1153df4ade7..37cd0de5d2a 100644
--- a/src/test/util/txmempool.cpp
+++ b/src/test/util/txmempool.cpp
@@ -150,7 +150,7 @@ void CheckMempoolV3Invariants(const CTxMemPool& tx_pool)
             // Check that special maximum virtual size is respected
             Assert(entry.GetTxSize() <= TRUC_MAX_VSIZE);
 
-            // Check that special v3 ancestor/descendant limits and rules are always
 respected
+            // Check that special TRUC ancestor/descendant limits and rules are alwa
ys respected
             Assert(entry.GetCountWithDescendants() <= TRUC_DESCENDANT_LIMIT);
             Assert(entry.GetCountWithAncestors() <= TRUC_ANCESTOR_LIMIT);
             Assert(entry.GetSizeWithDescendants() <= TRUC_MAX_VSIZE + TRUC_CHILD_MAX
_VSIZE);
@@ -159,12 +159,12 @@ void CheckMempoolV3Invariants(const CTxMemPool& tx_pool)
             // If this transaction has at least 1 ancestor, it's a "child" and has r
estricted weight.
             if (entry.GetCountWithAncestors() > 1) {
                 Assert(entry.GetTxSize() <= TRUC_CHILD_MAX_VSIZE);
-                // All v3 transactions must only have v3 unconfirmed parents.
+                // All TRUC transactions must only have TRUC unconfirmed parents.
                 const auto& parents = entry.GetMemPoolParentsConst();
                 Assert(parents.begin()->get().GetSharedTx()->version == TRUC_VERSION
);
             }
         } else if (entry.GetCountWithAncestors() > 1) {
-            // All non-v3 transactions must only have non-v3 unconfirmed parents.
+            // All non-TRUC transactions must only have non-TRUC unconfirmed parents
.
             for (const auto& parent : entry.GetMemPoolParentsConst()) {
                 Assert(parent.get().GetSharedTx()->version != TRUC_VERSION);
             }
diff --git a/src/test/util/txmempool.h b/src/test/util/txmempool.h
index 9986cabf7e9..57027cc0e62 100644
--- a/src/test/util/txmempool.h
+++ b/src/test/util/txmempool.h
@@ -47,12 +47,12 @@ std::optional<std::string>  CheckPackageMempoolAcceptResult(const
 Package& txns,
                                                             bool expect_valid,
                                                             const CTxMemPool* mempoo
l);
 
-/** For every transaction in tx_pool, check v3 invariants:
- * - a v3 tx's ancestor count must be within V3_ANCESTOR_LIMIT
- * - a v3 tx's descendant count must be within V3_DESCENDANT_LIMIT
- * - if a v3 tx has ancestors, its sigop-adjusted vsize must be within V3_CHILD_MA
X_VSIZE
- * - any non-v3 tx must only have non-v3 parents
- * - any v3 tx must only have v3 parents
+/** For every transaction in tx_pool, check TRUC invariants:
+ * - a TRUC tx's ancestor count must be within V3_ANCESTOR_LIMIT
+ * - a TRUC tx's descendant count must be within V3_DESCENDANT_LIMIT
+ * - if a TRUC tx has ancestors, its sigop-adjusted vsize must be within V3_CHILD_
MAX_VSIZE
+ * - any non-TRUC tx must only have non-TRUC parents
+ * - any TRUC tx must only have TRUC parents
  *   */
 void CheckMempoolV3Invariants(const CTxMemPool& tx_pool);
 
```

File: src/policy/truc_policy.h [hunk 4]
```diff
@@ -15,11 +15,12 @@
...
 #include <string>
```
Comment (ismaelsadeeq) [inline src/policy/truc_policy.h:17]: In f5b26e5b9be41cdab6e75aa0a3fbbb765a96ed06

This left out v3 mentions in the unit tests, maybe just leave out just the variable names as you stated from PR description.

File: src/policy/truc_policy.cpp [hunk 5]
```diff
@@ -14,7 +14,7 @@
 #include <numeric>
 #include <vector>
```
Comment (ismaelsadeeq) [inline src/policy/truc_policy.cpp:16]: Add `CheckMempoolV3Invariants` also?

File: test/functional/mempool_accept_v3.py [hunk 6]
```diff
@@ -644,4 +644,4 @@ def run_test(self):
```
Comment (ismaelsadeeq) [inline test/functional/mempool_accept_v3.py:645]: Maybe update `test/functional/wallet_create_tx.py` test.

see https://github.com/bitcoin/bitcoin/commit/d3fa531e67c0af77f52e754ad47a8396f2cab4c4, if it makes sense feel free to cherry-pick

File: src/policy/truc_policy.cpp (hunk 2 above)
Comment (glozow) [reply src/policy/truc_policy.cpp]: updated all

**Review by glozow — COMMENTED** [review]
thanks @ismaelsadeeq! Took your suggestions and fixed docs in a few more places

**Comment by instagibbs** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/30272/commits/926b8e39dcbc0a3a8a75ef0a29bdca2bf738d746

reviewed via "git range-diff master a6821a0 926b8e3"

**Review by ismaelsadeeq — APPROVED** [review]
Code review ACK 926b8e39dcbc0a3a8a75ef0a29bdca2bf738d746

**Comment by achow101** [comment]
ACK a6821a0a4e1cf2aba679962289f4beeefd3d539a

**Comment by achow101** [comment]
ACK 926b8e39dcbc0a3a8a75ef0a29bdca2bf738d746

**Comment by hebasto** [comment]
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.

**Pull request merged**